src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,560 @@
+{
+    static_cast<MulticastRoutingManager *>(aContext)->HandleBackboneMulticastListenerEvent(
+        aEvent, static_cast<const Ip6::Address &>(*aAddress));
+}
+
+void MulticastRoutingManager::HandleBackboneMulticastListenerEvent(otBackboneRouterMulticastListenerEvent aEvent,
+                                                                   const Ip6::Address &                   aAddress)
+{
+    switch (aEvent)
+    {
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ADDED:
+        Add(aAddress);
+        break;
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED:
+        Remove(aAddress);
+        break;
+    }
+}
+
+void MulticastRoutingManager::Enable(void)
+{
+    otError error = InitMulticastRouterSock();
+
+    otLogResultPlat(error, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Disable(void)
+{
+    FinalizeMulticastRouterSock();
+
+    otLogResultPlat(OT_ERROR_NONE, "MulticastRoutingManager: %s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Add(const Ip6::Address &aAddress)
+{
+    assert(mListenerSet.find(aAddress) == mListenerSet.end());
+    mListenerSet.insert(aAddress);
+
+    VerifyOrExit(IsEnabled());
+
+    UnblockInboundMulticastForwardingCache(aAddress);
+
+exit:
+    otLogResultPlat(OT_ERROR_NONE, "MulticastRoutingManager: %s: %s", __FUNCTION__, aAddress.ToString().AsCString());
+}
+
+void MulticastRoutingManager::Remove(const Ip6::Address &aAddress)
+{
+    otError error = OT_ERROR_NONE;
+
+    assert(mListenerSet.find(aAddress) != mListenerSet.end());
+    mListenerSet.erase(aAddress);
+
+    VerifyOrExit(IsEnabled());
+
+    RemoveInboundMulticastForwardingCache(aAddress);
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s: %s", __FUNCTION__, aAddress.ToString().AsCString());
+}
+
+void MulticastRoutingManager::UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd) const
+{
+    VerifyOrExit(IsEnabled());
+
+    FD_SET(mMulticastRouterSock, &aReadFdSet);
+    aMaxFd = std::max(aMaxFd, mMulticastRouterSock);
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Process(const fd_set &aReadFdSet)
+{
+    VerifyOrExit(IsEnabled());
+
+    ExpireMulticastForwardingCache();
+
+    if (FD_ISSET(mMulticastRouterSock, &aReadFdSet))
+    {
+        ProcessMulticastRouterMessages();
+    }
+
+exit:
+    return;
+}
+
+otError MulticastRoutingManager::InitMulticastRouterSock(void)
+{
+    otError             error = OT_ERROR_NONE;
+    int                 one   = 1;
+    struct icmp6_filter filter;
+    struct mif6ctl      mif6ctl;
+
+    VerifyOrExit(!IsEnabled());
+
+    // Create a Multicast Routing socket
+    mMulticastRouterSock = socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6);
+    VerifyOrExit(mMulticastRouterSock != -1, error = OT_ERROR_FAILED);
+
+    // Enable Multicast Forwarding in Kernel
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_INIT, &one, sizeof(one)),
```

Comment:
VerifyOrDie? It seems to be a critical error.

Reply:
It's arguable to some extent. One may argue that Backbone Router without Multicast routing is also useful, thus should keep running. 
But agree that VerifyOrDie should be better for disclosing issues. Will change.